Require fsspec URLs, refactor

brews opened the pull request:
With this PR just about all data inputs need to be `fsspec` URLs. All storage authentication gets passed through fsspec-storage backend default environment variables. `dodola.services` functions no longer require the `storage` arg. 

In short, we spec storage with URLs like:
```
dodola cleancmip6 az://raw/gfdl.zarr az://clean/gfdl.zarr
```

Storage is authenticated completely through input `fsspec` URLs on a data-by-data basis. This means functions requiring multiple input data can read the data from different `fsspec` backends.

In terms of package architecture, storage I/O (`dodola/repository.py`) would now happen in `dodola/services.py`. `dodola/core.py` and notably `dodola/cli.py`. Both are now decoupled from storage logic.

`dodola` will accept URLs for S3 (via `s3fs`) and GCS buckets (via `gcsfs`), among others. See `fsspec` registry for list of supported protocols and the corresponding required packages. https://filesystem-spec.readthedocs.io/en/latest/_modules/fsspec/registry.html

Because we're mostly using Azure storage on the backend for this project, users won't be able to pass in credentials with the old `AZURE_STORAGE_ACCOUNT` and `AZURE_STORAGE_KEY`, for example. Under this PR, users need the `AZURE_STORAGE_ACCOUNT_NAME` and `AZURE_STORAGE_ACCOUNT_KEY` because these are the environment variables used by the `fsspec` backend for Azure Gen 2 Datalake, `adlfs`. This is another breaking change.

There was some light refactoring to accommodate all this.

 Close #56, close #55

dgergel commented on dodola/repository.py:
```diff
@@ -2,142 +2,58 @@
-        Mapping of `{path: data}` to immediately store into memory filesystem.
+    root : str
+    check : bool
+        Do touch at storage to check for write access.
```
I'm a little unclear on what the description of `check` means here...

brews replied:
This was removed all together in PR #60. It's no longer an issue. Marking as resolved.

dgergel commented on dodola/tests/test_repository.py:
```diff
@@ -1,15 +1,16 @@
-    storage = memory_repository(storage={"foo": Dataset({"bar": 123})})
-    assert storage.read(url_or_path="foo") == Dataset({"bar": 123})
+    url = "memory://test_memory_repository_read.zarr"
+    Dataset({"bar": 321}).to_zarr(url)  # Manually write to memory FS.
```
Interesting! Thought you had to import `Dataset` from `xarray.core`.

brews replied:
That's where it might be defined etc but they import it in `xarray.__init__.py` from `xarray.core` for a nice interface kinda like `np.array` and `pd.DataFrame` you have `xr.Dataset` and `xr.DataArray`!

dgergel commented on environment.yaml:
```diff
@@ -6,11 +6,13 @@ dependencies:
 - pytest
 - python=3.8
 - rechunker
+- s3fs
```
Super cool how platform-agnostic this is now!

brews replied:
Yeah, it's platform-agnostic (as long as you're in S3, GCS, or AZ). I hope it works.

dgergel reviewed (changes requested):
@brews this is really awesome, thanks for putting this together! Very excited about where this is all going.

brews commented on dodola/repository.py:
```diff
@@ -2,142 +2,58 @@
-    This is commonly instantiated through ``adl_repository``, or for testing
-    with ``memory_repository``.
+def read(url_or_path):
+    """Read Dataset from Zarr store
```
Ehhhhh, I'm kinda on the line about this one. In this case, I didn't originally put "zarr" because the fact that we're writing to zarr is an implementation detail that I'm trying to abstract away.

I'd prefer to keep it as is for this PR because it matches the interface we refactored from. Marking this resolved.

brews commented on dodola/repository.py:
```diff
@@ -2,142 +2,58 @@
-    and `tenant_id` for authentication with service principal
-    credentials. Initializing arguments are passed to ``adlfs.AzureBlobFileSystem``.
+    This opens Zarr store with mode "w" and is called with with
+    ``compute=True``, so any lazy computations will be completed.
```
Thank you! Fixed with `d8875acf`.

brews commented:
Hey, @dgergel  I think this is ready for re-review as soon as you have a second.

I merged in the changes from the rechunker rewrite (PR #60 )and addressed issues from previous review. Check a look! Thanks!

dgergel reviewed (approved):
LGTM, thanks @brews!

brews commented:
Thanks so much for the help cleaning this up, @dgergel !

I'm merging this.
